File: api/routers/hygiene.py
import csv, io
from datetime import datetime
from typing import Optional
from fastapi import APIRouter, File, HTTPException, UploadFile
from fastapi.responses import StreamingResponse
from pydantic import BaseModel
from ..services.deduplicator import deduplicate, detect_role_emails
from ..services.job_store import create_job, update_job, append_result, get_all_results, get_job

router = APIRouter(prefix="/hygiene", tags=["Hygiene"])

@router.post("/clean")
async def clean_list(file: UploadFile = File(...)):
    if not file.filename or not file.filename.endswith(".csv"):
        raise HTTPException(400, "Only .csv files accepted.")
    raw  = await file.read()
    text = raw.decode("utf-8-sig") if raw[:3] == b'\xef\xbb\xbf' else raw.decode("latin-1")
    reader = csv.DictReader(io.StringIO(text))
    rows   = [dict(row) for row in reader]
    total  = len(rows)
    if total == 0: raise HTTPException(400, "CSV is empty.")

    # 1. Deduplicate
    rows, dupes = deduplicate(rows)
    # 2. Role-based detection
    rows, role_count = detect_role_emails(rows)
    # 3. Filter out role-based from clean list
    clean = [r for r in rows if not r.get("_role_based")]
    # Clean up internal flags
    for r in clean:
        r.pop("_role_based", None)

    job_id = await create_job(total, job_type="hygiene")
    for r in clean:
        await append_result(job_id, r)
    await update_job(job_id, status="completed", completed_at=datetime.utcnow().isoformat(),
                     processed_rows=total, successful_rows=len(clean),
                     failed_rows=total - len(clean))

    return {
        "job_id":     job_id,
        "total":      total,
        "valid":      len(clean),
        "invalid":    total - len(clean),
        "duplicates": dupes,
        "role_based": role_count,
        "suppressed": 0,
    }

@router.get("/jobs/{job_id}/download")
async def download_clean(job_id: str):
    job = await get_job(job_id)
    if not job: raise HTTPException(404, "Job not found.")
    if job["status"] != "completed": raise HTTPException(409, "Not complete.")
    results = await get_all_results(job_id)
    if not results: raise HTTPException(404, "No results.")
    fieldnames = list(results[0].keys())
    def _gen():
        buf = io.StringIO()
        w   = csv.DictWriter(buf, fieldnames=fieldnames, extrasaction="ignore")
        w.writeheader(); yield buf.getvalue()
        for row in results:
            buf = io.StringIO()
            w.writerow(row); yield buf.getvalue()
    return StreamingResponse(_gen(), media_type="text/csv",
        headers={"Content-Disposition": f"attachment; filename=clean_{job_id}.csv"})

@router.post("/suppress")
async def add_suppression(email: str):
    from ..services.redis import redis_set
    await redis_set(f"suppressed:{email.lower().strip()}", True, ttl=365*24*3600)
    return {"suppressed": email}

@router.get("/suppression-list")
async def get_suppression_list():
    from ..services.redis import get_redis
    r    = await get_redis()
    keys = await r.keys("suppressed:*")
    return {"emails": [k.replace("suppressed:", "") for k in keys], "total": len(keys)}

File: api/routers/hygiene.js
import { Router } from "express";
import multer from "multer";
import { parse } from "csv-parse/sync";
import { stringify } from "csv-stringify/sync";
import { deduplicate, detectRoleEmails } from "../services/deduplicator.js";
import { createJob, updateJob, appendResult, getAllResults, getJob } from "../services/jobStore.js";
import { getRedis, redisSet, redisKeys } from "../services/redis.js";

export const hygieneRouter = Router();
const upload = multer({ storage: multer.memoryStorage() });

hygieneRouter.post("/hygiene/clean", upload.single("file"), async (req, res) => {
  if (!req.file) return res.status(400).json({ error: "No file uploaded." });
  let records;
  try {
    records = parse(req.file.buffer.toString("utf-8"), { columns: true, skip_empty_lines: true, trim: true });
  } catch (e) {
    return res.status(400).json({ error: `CSV parse error: ${e.message}` });
  }
  if (!records.length) return res.status(400).json({ error: "CSV is empty." });
  const total = records.length;
  const { clean: deduped, dupes } = deduplicate(records);
  const { rows: annotated, roleCount } = detectRoleEmails(deduped);
  const clean = annotated.filter(r => !r._role_based).map(({ _role_based, ...r }) => r);
  const jobId = await createJob(total, "hygiene");
  for (const r of clean) await appendResult(jobId, r);
  await updateJob(jobId, {
    status: "completed", completed_at: new Date().toISOString(),
    processed_rows: total, successful_rows: clean.length,
    failed_rows: total - clean.length,
  });
  res.json({
    job_id: jobId, total, valid: clean.length,
    invalid: total - clean.length,
    duplicates: dupes, role_based: roleCount, suppressed: 0,
  });
});

hygieneRouter.get("/hygiene/jobs/:jobId/download", async (req, res) => {
  const job = await getJob(req.params.jobId);
  if (!job) return res.status(404).json({ error: "Job not found." });
  if (job.status !== "completed") return res.status(409).json({ error: "Not complete." });
  const results = await getAllResults(req.params.jobId);
  if (!results.length) return res.status(404).json({ error: "No results." });
  const csv = stringify(results, { header: true, columns: Object.keys(results[0]) });
  res.setHeader("Content-Type", "text/csv");
  res.setHeader("Content-Disposition", `attachment; filename=clean_${req.params.jobId}.csv`);
  res.send(csv);
});

hygieneRouter.post("/hygiene/suppress", async (req, res) => {
  const { email } = req.body;
  if (!email) return res.status(422).json({ error: "email is required." });
  await redisSet(`suppressed:${email.toLowerCase().trim()}`, true, 365 * 24 * 3600);
  res.json({ suppressed: email });
});

hygieneRouter.get("/hygiene/suppression-list", async (req, res) => {
  const keys   = await redisKeys("suppressed:*");
  const emails = keys.map(k => k.replace("suppressed:", ""));
  res.json({ emails, total: emails.length });
});
